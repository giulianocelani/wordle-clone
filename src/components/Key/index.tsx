/* eslint-disable default-case */
import { mappings } from '../../constants';
import useStore from '../../store';
import { LetterState } from '../../utils';

type IProps = {
	letter: string;
	onClick: (letter: string) => void;
};

const getKeyStyle = (
	letter: string,
	keysPressed: Record<string, LetterState>
) => {
	const styles: string[] = [];

	if (keysPressed[letter]) {
		styles.push('text-white');
		if (keysPressed[letter] === LetterState.MISS) {
			styles.push('hover:cursor-not-allowed');
		}
	} else {
		styles.push('border hover:cursor-pointer hover:border-gray-400');
	}

	styles.push(mappings.StateToColor[keysPressed[letter]]);
	return styles.join(' ');
};

const Key = ({ letter, onClick }: IProps) => {
	const { keysPressed } = useStore();

	const handleOnClick = () => {
		if (typeof letter === 'string') {
			onClick(letter);
		} else {
			onClick('Backspace');
		}
	};

	return (
		<button
			type='button'
			onClick={handleOnClick}
			className={`flex items-center justify-center p-2 sm:p-3 md:p-5 max-h-12 ${
				letter.length > 1 ? 'min-w-max' : 'min-w-[2rem]'
			} rounded-md text-md lg:text-xl font-semibold transition-colors ${getKeyStyle(
				letter,
				keysPressed
			)}`}
			disabled={keysPressed[letter] === LetterState.MISS}
		>
			{letter === 'Backspace' ? (
				<svg
					xmlns='http://www.w3.org/2000/svg'
					className='h-5'
					viewBox='0 0 20 20'
					fill='currentColor'
				>
					<path
						fillRule='evenodd'
						d='M6.707 4.879A3 3 0 018.828 4H15a3 3 0 013 3v6a3 3 0 01-3 3H8.828a3 3 0 01-2.12-.879l-4.415-4.414a1 1 0 010-1.414l4.414-4.414zm4 2.414a1 1 0 00-1.414 1.414L10.586 10l-1.293 1.293a1 1 0 101.414 1.414L12 11.414l1.293 1.293a1 1 0 001.414-1.414L13.414 10l1.293-1.293a1 1 0 00-1.414-1.414L12 8.586l-1.293-1.293z'
						clipRule='evenodd'
					/>
				</svg>
			) : (
				letter.toUpperCase()
			)}
		</button>
	);
};

export default Key;
