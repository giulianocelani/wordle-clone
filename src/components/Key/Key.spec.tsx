import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Key from '.';

describe('Key', () => {
	let wrapper = shallow(<Key letter='A' onClick={jest.fn()} />);

	it('renders correctly with single character', () => {
		expect(toJson(wrapper, { mode: 'deep' })).toMatchSnapshot();
		expect(wrapper.find('button').length).toBe(1);

		expect(wrapper.find('svg').length).toBe(0);
	});

	it('renders correct with multiple characters', () => {
		wrapper = shallow(<Key letter='Enter' onClick={jest.fn()} />);
		expect(toJson(wrapper, { mode: 'deep' })).toMatchSnapshot();
		expect(wrapper.find('button').length).toBe(1);

		expect(wrapper.find('svg').length).toBe(0);
	});

	it('renders correct with svg', () => {
		wrapper = shallow(<Key letter='Backspace' onClick={jest.fn()} />);
		expect(toJson(wrapper, { mode: 'deep' })).toMatchSnapshot();

		expect(wrapper.find('button').length).toBe(1);

		expect(wrapper.find('svg').length).toBe(1);
	});
});
