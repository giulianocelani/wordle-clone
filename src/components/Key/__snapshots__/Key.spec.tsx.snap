// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Key renders correct with multiple characters 1`] = `
<button
  className="flex items-center justify-center p-2 sm:p-3 md:p-5 max-h-12 min-w-max rounded-md text-md lg:text-xl font-semibold transition-colors border hover:cursor-pointer hover:border-gray-400 "
  disabled={false}
  onClick={[Function]}
  type="button"
>
  ENTER
</button>
`;

exports[`Key renders correct with svg 1`] = `
<button
  className="flex items-center justify-center p-2 sm:p-3 md:p-5 max-h-12 min-w-max rounded-md text-md lg:text-xl font-semibold transition-colors border hover:cursor-pointer hover:border-gray-400 "
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <svg
    className="h-5"
    fill="currentColor"
    viewBox="0 0 20 20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M6.707 4.879A3 3 0 018.828 4H15a3 3 0 013 3v6a3 3 0 01-3 3H8.828a3 3 0 01-2.12-.879l-4.415-4.414a1 1 0 010-1.414l4.414-4.414zm4 2.414a1 1 0 00-1.414 1.414L10.586 10l-1.293 1.293a1 1 0 101.414 1.414L12 11.414l1.293 1.293a1 1 0 001.414-1.414L13.414 10l1.293-1.293a1 1 0 00-1.414-1.414L12 8.586l-1.293-1.293z"
      fillRule="evenodd"
    />
  </svg>
</button>
`;

exports[`Key renders correctly with single character 1`] = `
<button
  className="flex items-center justify-center p-2 sm:p-3 md:p-5 max-h-12 min-w-[2rem] rounded-md text-md lg:text-xl font-semibold transition-colors border hover:cursor-pointer hover:border-gray-400 "
  disabled={false}
  onClick={[Function]}
  type="button"
>
  A
</button>
`;
